File: Assets/src/model/tasks/GenericTask.cs
using System;
using System.Collections.Generic;

/// Close-ended
[System.Serializable]
public class GenericTask : DoOnceTask {
  private ActionType _type;
  ActionType type => _type;
  /// TODO-SERIALIZE audit all Actions
  public GenericTask(Actor actor, Action<Actor> action, ActionType type = ActionType.GENERIC) : base(actor) {
    _type = type;
    Action = action;
    Name = Util.WithSpaces(Action.Method.Name);
  }

  public Action<Actor> Action { get; }

  protected override BaseAction GetNextActionImpl() {
    return new GenericBaseAction(actor, Action);
  }
}

File: web/src/model/tasks/GenericTask.ts
import { DoOnceTask } from '../ActorTask';
import { GenericBaseAction, type BaseAction } from '../BaseAction';
import type { Actor } from '../Actor';
import type { Player } from '../Player';

export class GenericTask extends DoOnceTask {
  readonly action: () => void;

  constructor(actor: Actor, action: () => void) {
    super(actor);
    this.action = action;
  }

  protected getNextActionImpl(): BaseAction {
    return new GenericBaseAction(this.actor, this.action);
  }
}

export class GenericPlayerTask extends GenericTask {
  constructor(actor: Player, action: () => void) {
    super(actor, action);
  }
}
